### PR TITLE
[FIXED JENKINS-43199] Defensively ensure the root stream is always closed even when plugins misbehave

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-43199](https://issues.jenkins-ci.org/browse/JENKINS-43199) and [JENKIN-45057](https://issues.jenkins-ci.org/browse/JENKINS-45057)

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/Run.java [hunk 1]
```diff
@@ -1801,6 +1801,7 @@ protected final void execute(@Nonnull RunExecution job) {
 		LOGGER.log(Level.SEVERE, "Failed to rotate log",e);
 	    }
         } finally {
+            IOUtils.closeQuietly(listener);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Run.java]: Logging would be useful here. IIRC there is a method in hudson.Util or other util class

File: core/src/main/java/hudson/model/StreamBuildListener.java [hunk 2]
```diff
@@ -40,16 +41,26 @@
  * @author Kohsuke Kawaguchi
  */
 public class StreamBuildListener extends StreamTaskListener implements BuildListener {
+    private final Closeable[] additionalClosables;
+
+    public StreamBuildListener(OutputStream out, Charset charset, Closeable... additionalClosables) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/StreamBuildListener.java]: Add some javadoc?

File: core/src/main/java/hudson/model/StreamBuildListener.java [hunk 3]
```diff
@@ -80,5 +93,41 @@ public void finished(Result result) {
         getLogger().println("Finished: "+result);
     }
 
+    @Override
+    public void close() throws IOException {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/StreamBuildListener.java]: It implies that all classes are tolerant against douple close() call. They should according to the Closeable speck, so fine for me

**Review by oleg-nenashev — COMMENTED** [review]
Mostly LGTM, minor comments

File: core/src/main/java/hudson/model/Run.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Run.java]: This is already called at line 1788, so what is the effect of calling it again?

File: core/src/main/java/hudson/model/StreamBuildListener.java [hunk 4]
```diff
@@ -40,16 +41,26 @@
  * @author Kohsuke Kawaguchi
  */
 public class StreamBuildListener extends StreamTaskListener implements BuildListener {
+    private final Closeable[] additionalClosables;
```
Comment (jglick) [inline core/src/main/java/hudson/model/StreamBuildListener.java]: This new API seems unnecessary complicated. Would suffice for `RunExecution.execute` to retain the `OutputStream logger` as a separate local variable, and close _that_ in its `finally` block.

**Review by jglick — COMMENTED** [review]
Should work, though #2954 would accomplish the same with no new API and less fuss.

File: core/src/main/java/hudson/model/StreamBuildListener.java [hunk 5]
```diff
@@ -80,5 +93,41 @@ public void finished(Result result) {
...
+    public void close() throws IOException {
+        try {
+            super.close();
+        } catch (IOException e) {
+            if (additionalClosables != null) {
+                for (Closeable c: additionalClosables) {
+                    try {
+                        c.close();
+                    } catch (IOException e1) {
+                        e.addSuppressed(e1);
```
Comment (jglick) [inline core/src/main/java/hudson/model/StreamBuildListener.java]: If we are to do something this fancy, I think it should be a more general-purpose utility, as proposed here: https://github.com/jenkinsci/remoting/pull/104#r75861893

File: core/src/main/java/hudson/model/StreamBuildListener.java (hunk 4 above)
Comment (jglick) [inline core/src/main/java/hudson/model/StreamBuildListener.java]: Making this field be non`transient` seems to cause [many legitimate-looking test failures](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-2953/1/testReport/).

**Review by jglick — CHANGES_REQUESTED** [review]

**Comment by stephenc** [comment]
Closing in favour of #2954

**Pull request closed without merging**